### ISSUE TITLE
 problem about inferrence

### ISSUE THREAD

**Issue by reclusezz** [open-mmlab/mmclassification#424]
when i infer my model, i meet this problem.
Via the inferrence file of mmclassification, it turns out that the label of the image is 1 and the pred score is 0.8889318704605103.
but when i use my own inferrence file, the output of fc is tensor([[ 0.5133, -0.5640]], device='cuda:0').
Do you have an ideas where i have operate incorrectly?

my own inferrence file is as below:
```
def forward_pytorch(weightfile, image):
    net=resnet.resnet18(num_classes=2)
    checkpoint = torch.load(weightfile)
    new_state_dict = OrderedDict()
    for k, v in checkpoint.items():
        if k == "state_dict":
            for layer_name, weight in v.items():
                layer_name1 = layer_name.replace("backbone.", "")
                layer_name1 = layer_name1.replace("head.", "")
                print(layer_name, layer_name1)
                new_state_dict[layer_name1] = weight
    net.load_state_dict(new_state_dict)
    net.cuda()
    net.eval()
    normalize = transforms.Normalize(mean=[123.675, 116.28, 103.53], std=[58.395, 57.12, 57.375])
    transform1 = transforms.Compose([
        transforms.ToTensor(),
        normalize
        ]
    )
    image = transform1(image)  
    image = Variable(image.cuda())
    image = Variable(torch.unsqueeze(image, dim=0).float(), requires_grad=False)
    with torch.no_grad():
        blobs = net.forward(image)
    return blobs
```

**Comment by Ezra-Yu** [comment]
Those points you should notice:
1. image channel before transform must be in "RGB" order in [mmcls config resnet18.](https://github.com/open-mmlab/mmclassification/blob/6a0a76af0c176f381c8534097333f37a9e65a6bf/configs/_base_/datasets/imagenet_bs32.py#L4)
2. Whether all the parameters in the weight are loaded correctly.
3. there is [Resize](https://github.com/open-mmlab/mmclassification/blob/6a0a76af0c176f381c8534097333f37a9e65a6bf/configs/_base_/datasets/imagenet_bs32.py#L16) data augment before forward.
4. the ["ImageToTensor"](https://github.com/open-mmlab/mmclassification/blob/6a0a76af0c176f381c8534097333f37a9e65a6bf/configs/_base_/datasets/imagenet_bs32.py#L19) in pipeline have a op of [img.transpose](https://github.com/open-mmlab/mmclassification/blob/6a0a76af0c176f381c8534097333f37a9e65a6bf/mmcls/datasets/pipelines/formating.py#L62); transpose the input(1, 224, 224, 3) to (1, 3, 224, 224)

**Comment by reclusezz** [comment]
hello， i load the image by PIL, it is supposed to be RGB. Meanwhile, i do resize the image before foward and check the input(1, 3, 224, 224). The way of loading model parameters seems not to be wrong. 

this is is my file:
def load_image_pytorch(imgfile):  
    img = Image.open(imgfile)
    img = img.resize((224, 224))
    return img


this is the ouput:torch.Size([1, 3, 224, 224])
pytorch forward time %d 0.006729602813720703
tensor([[ 0.5224, -0.5617]], device='cuda:0')
label: 0


the problem really DO troubles me a lot

**Comment by Ezra-Yu** [comment]
We don't have your picture, code and weight, so we don't see the problem for the time being. You can eliminate the problem item by item,
1. Check the model, use a fixed random (1, 3, 224, 224) tensor, go through the two models respectively, check the difference, if there is no difference, then it is the problem of data load
2. After ensuring that the model is ok, check the data problem; the op resize in mmcls is different from your resize, which is more complicated; or you can change the Resize in config to your own resize; or delete Resize in config, directly Use (1, 3, w, h) for reasoning, and delete the resize in your code accordingly. See if your output is suppressed.


this a demo modified from [./demo/image_demo.py](https://github.com/open-mmlab/mmclassification/blob/master/demo/image_demo.py); This way you can get the same tensor by fixing a random seed.
```
def main():
    parser = ArgumentParser()
    parser.add_argument('config', help='Config file')
    parser.add_argument('checkpoint', help='Checkpoint file')
    parser.add_argument(
        '--device', default='cpu', help='Device used for inference')
    args = parser.parse_args()

    # build the model from a config file and a checkpoint file
    model = init_model(args.config, args.checkpoint, device=args.device)
    # train a single image
    from mmcv.runner.utils import set_random_seed
    set_random_seed(100)
    img = torch.randn( (1, 3, 224 ,224) )
      with torch.no_grad():
          scores = model(img, return_loss=False)
          pred_score = np.max(scores, axis=1)[0]
          pred_label = np.argmax(scores, axis=1)[0]
          result = {'pred_label': pred_label, 'pred_score': float(pred_score)}


```

**Comment by reclusezz** [comment]
Thx！Following your advice, I have found that this problem is mainly caused by data aug. The mean and std of Normalization in mmclassification resnet18 is [123.675, 116.28, 103.53] and [58.395, 57.12, 57.375], respectively. It seems that Normalization op is done before ImageToTensor. Is there any difference if I do Normalization op after ImageToTensor with mean and std ranging from 0 to 1?

**Comment by Ezra-Yu** [comment]
No different. But you have to make sure that image go through ["ImageToTensor", "Normalization"] or  [ "Normalization", "ImageToTensor"] get the same tensor.

To make things simple, you can use a simple input image as example , such a (2, 2, 3) np.array as image input.
the mmcls pipeline can use as below:
```
import numpy as np
from mmcls.datasets.pipelines import Compose

NORM_MEAN = [103.53, 116.28, 123.675]
NORM_STD = [57.375, 57.12, 58.395]
pipeline_cfg = [
    dict(type='Normalize', mean=NORM_MEAN, std=NORM_STD, to_rgb=False),
    dict(type='ImageToTensor', keys=['img']),
]
pipeline = Compose(pipeline_cfg)
input = {'img':np.ones( [2, 2, 3])}
output = pipeline(input)

```

then, campare with your pipeline.

**Comment by reclusezz** [comment]
I tried using a (2, 2, 3) np.array as image input. Method A and B got the same output:


Method A:
input = np.ones( [224, 224, 3] )
NORM_MEAN = [123.675, 116.28, 103.53]
NORM_STD = [58.395, 57.12, 57.375]
pipeline_cfg = [
    dict(type='Normalize', mean=NORM_MEAN, std=NORM_STD, to_rgb=True),
    dict(type='ImageToTensor', keys=['img']),
]
pipeline = Compose(pipeline_cfg)
output = pipeline(input)


Method B:
input = np.ones( [224, 224, 3] )
transform1 = transforms.Compose([
    transforms.ToTensor(),
    transforms.Normalize(mean=[123.675, 116.28, 103.53], std=[58.395, 57.12, 57.375]) 
    ])
output = transform1(input)

Output:
{'img': tensor([[[-2.1008, -2.1008],
         [-2.1008, -2.1008]],

        [[-2.0182, -2.0182],
         [-2.0182, -2.0182]],

        [[-1.7870, -1.7870],
         [-1.7870, -1.7870]]]), 'img_norm_cfg': {'mean': array([123.675, 116.28 , 103.53 ], dtype=float32), 'std': array([58.395, 57.12 , 57.375], dtype=float32), 'to_rgb': True}}


However, when it comes to a real picture:

Method A:
img = cv2.imread(picpath)
input = {'img':img}
NORM_MEAN = [103.53, 116.28, 123.675]
NORM_STD = [57.375, 57.12, 58.395]
pipeline_cfg = [
    dict(type='Normalize', mean=NORM_MEAN, std=NORM_STD, to_rgb=False),
    dict(type='ImageToTensor', keys=['img']),
]
pipeline = Compose(pipeline_cfg)
output = pipeline(input)
print(output)

Output A:
{'img': tensor([[[-1.1770, -1.2467, -1.3513,  ..., -1.7347, -1.6999, -1.7347],
         [-1.0898, -1.2467, -1.3164,  ..., -1.6824, -1.6999, -1.6999],
         [-1.1421, -1.1944, -1.1596,  ..., -1.6999, -1.7347, -1.6999],
         ...,
         [-0.8807, -0.8807, -0.8807,  ..., -0.7587, -0.7587, -0.7587],
         [-0.8633, -0.8807, -0.8807,  ..., -0.7238, -0.7413, -0.7238],
         [-0.8458, -0.8458, -0.8458,  ..., -0.7238, -0.7413, -0.7238]],

        [[-1.2654, -1.3354, -1.4405,  ..., -1.8957, -1.8606, -1.8957],
         [-1.1779, -1.3354, -1.4055,  ..., -1.8431, -1.8606, -1.8606],
         [-1.2304, -1.2829, -1.2479,  ..., -1.8606, -1.8957, -1.8606],
         ...,
         [-0.9503, -0.9503, -0.9503,  ..., -0.8627, -0.8803, -0.8803],
         [-0.9328, -0.9503, -0.9678,  ..., -0.8452, -0.8803, -0.8627],
         [-0.9153, -0.9153, -0.9328,  ..., -0.8452, -0.8803, -0.8627]],

        [[-1.3815, -1.4500, -1.5528,  ..., -2.0665, -2.0323, -2.0665],
         [-1.2959, -1.4500, -1.5185,  ..., -2.0152, -2.0323, -2.0323],
         [-1.3302, -1.3815, -1.3473,  ..., -2.0323, -2.0665, -2.0323],
         ...,
         [-1.1075, -1.1075, -1.1075,  ..., -1.0219, -1.0048, -1.0048],
         [-1.0904, -1.1075, -1.0904,  ..., -1.0390, -1.0219, -1.0048],
         [-1.0733, -1.0733, -1.0562,  ..., -1.0390, -1.0219, -1.0048]]]), 'img_norm_cfg': {'mean': array([103.53 , 116.28 , 123.675], dtype=float32), 'std': array([57.375, 57.12 , 58.395], dtype=float32), 'to_rgb': False}}



Method B:
input = cv2.imread(picpath)
transform1 = transforms.Compose([
    transforms.ToTensor(), 
    transforms.Normalize(mean=[123.675, 116.28, 103.53], std=[58.395, 57.12, 57.375]) 
    ])
output = transform1(input)
print(output)

Output B:
tensor([[[-2.1155, -2.1158, -2.1162,  ..., -2.1176, -2.1175, -2.1176],
         [-2.1152, -2.1158, -2.1160,  ..., -2.1174, -2.1175, -2.1175],
         [-2.1154, -2.1156, -2.1154,  ..., -2.1175, -2.1176, -2.1175],
         ...,
         [-2.1143, -2.1143, -2.1143,  ..., -2.1139, -2.1139, -2.1139],
         [-2.1143, -2.1143, -2.1143,  ..., -2.1137, -2.1138, -2.1137],
         [-2.1142, -2.1142, -2.1142,  ..., -2.1137, -2.1138, -2.1137]],

        [[-2.0327, -2.0330, -2.0334,  ..., -2.0352, -2.0350, -2.0352],
         [-2.0324, -2.0330, -2.0332,  ..., -2.0350, -2.0350, -2.0350],
         [-2.0326, -2.0328, -2.0326,  ..., -2.0350, -2.0352, -2.0350],
         ...,
         [-2.0315, -2.0315, -2.0315,  ..., -2.0311, -2.0312, -2.0312],
         [-2.0314, -2.0315, -2.0315,  ..., -2.0310, -2.0312, -2.0311],
         [-2.0313, -2.0313, -2.0314,  ..., -2.0310, -2.0312, -2.0311]],

        [[-1.8015, -1.8018, -1.8022,  ..., -1.8042, -1.8041, -1.8042],
         [-1.8012, -1.8018, -1.8021,  ..., -1.8040, -1.8041, -1.8041],
         [-1.8013, -1.8015, -1.8014,  ..., -1.8041, -1.8042, -1.8041],
         ...,
         [-1.8004, -1.8004, -1.8004,  ..., -1.8001, -1.8000, -1.8000],
         [-1.8003, -1.8004, -1.8003,  ..., -1.8001, -1.8001, -1.8000],
         [-1.8003, -1.8003, -1.8002,  ..., -1.8001, -1.8001, -1.8000]]])


The output of A and B seems not to be the same.

**Comment by Ezra-Yu** [comment]
Do not use np.ones();
Try a self-designed array or a randn.

**Comment by Ezra-Yu** [comment]
I think is ` dict(type='ImageToTensor', keys=['img'])`changes image channel order.

**Comment by reclusezz** [comment]
dict(type='ImageToTensor', keys=['img']) would change the image channel order , and transforms.ToTensor() will do the same.

**Comment by Ezra-Yu** [comment]
`dict(type='Normalize', mean=NORM_MEAN, std=NORM_STD, to_rgb=True)`; the `to_rgb=True` will change a img from RGB to BGR or a BGR to a RBG; np.ones can not find the oreder problem. The NORM_MEAN and NORM_STD has its own RGB oreder.

in my script, everything work well.
![image](https://user-images.githubusercontent.com/18586273/131513847-efb1dcee-b080-4275-8679-2df871b7099c.png)

**Comment by reclusezz** [comment]
Thx! I have found the key. When I change the type of the array loaded from a real image to  float32, the output tensor of the two method will be the same.

**Comment by Ezra-Yu** [comment]
you are welcome.
